### PR TITLE
[FLINK-20840][table-planner] Properly transpose projection to join

### PR DESCRIPTION
**Pull request by leonardBang**
## What is the purpose of the change

* Currently the rule `FlinkProjectJoinTransposeRule` was added in logical optimize phase which uses volcano optimizer, the optimized plan is uncertain because it uses CBO model.

## Brief change log

  - Add a `PROJECT_REWRITE` phase in hep optimizer which uses RBO model.

## Verifying this change

* The tests in TemporalJoinTest/LookupJoinTest cover this change

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by leonardBang** [comment]
cc： @godfreyhe could you help review this PR?

**Comment by leonardBang** [comment]
@flinkbot run azure

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java [hunk 1]
```diff
@@ -308,6 +308,14 @@
     // ------------------------------------------------------------------------
     //  Other Exec Options
     // ------------------------------------------------------------------------
+    @Documentation.TableOption(execMode = Documentation.ExecMode.BATCH_STREAMING)
```
Comment (godfreyhe) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/ExecutionConfigOptions.java:311]: why we need to introduce this ?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/delegation/PlannerContext.java [hunk 2]
```diff
@@ -83,8 +83,8 @@
 @Internal
 public class PlannerContext {
 
-    private final RelDataTypeSystem typeSystem = new FlinkTypeSystem();
-    private final FlinkTypeFactory typeFactory = new FlinkTypeFactory(typeSystem);
+    private final RelDataTypeSystem typeSystem;
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/delegation/PlannerContext.java:86]: does this really need to change ?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/FlinkTypeFactory.scala [hunk 3]
```diff
@@ -160,6 +160,14 @@ class FlinkTypeFactory(typeSystem: RelDataTypeSystem)
           case TimestampKind.ROWTIME => createRowtimeIndicatorType(true)
           case TimestampKind.REGULAR => createSqlType(TIMESTAMP, timestampType.getPrecision)
         }
+      case LogicalTypeRoot.TIMESTAMP_WITH_LOCAL_TIME_ZONE =>
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/FlinkTypeFactory.scala:163]: ditto

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java [hunk 4]
```diff
@@ -18,41 +18,19 @@
...
+
+import org.apache.calcite.sql.*;
+import org.apache.calcite.sql.fun.*;
+import org.apache.calcite.sql.type.*;
+import org.apache.calcite.sql.util.*;
+import org.apache.calcite.sql.validate.*;
+
+import java.util.*;
+
+import static org.apache.flink.table.planner.plan.type.FlinkReturnTypes.*;
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java:33]: ditto ?

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution @leonardBang , many changes are unrelated in this pr?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/delegation/PlannerContext.java (hunk 2 above)
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/delegation/PlannerContext.java:86]: Oops，they should only appear in my another developing branch.

File: flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/stream/sql/MiniBatchIntervalInferTest.xml [hunk 5]
```diff
@@ -290,7 +290,7 @@ LogicalProject(a=[$0], b=[$1])
     <Resource name="optimized exec plan">
       <![CDATA[
 Calc(select=[a, b])
-+- Join(joinType=[InnerJoin], where=[(a0 = a1)], select=[a, b, a0, b0, a1], leftInputSpec=[JoinKeyContainsUniqueKey], rightInputSpec=[HasUniqueKey])
++- Join(joinType=[InnerJoin], where=[(a0 = a1)], select=[a, b, a0, a1], leftInputSpec=[JoinKeyContainsUniqueKey], rightInputSpec=[NoUniqueKey])
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/stream/sql/MiniBatchIntervalInferTest.xml:293]: nit: it's may be a bad case (depend on input data)

**Review by godfreyhe — APPROVED** [review]
LGTM

**Pull request merged**